### PR TITLE
[ci] Run all archs and full tests for CI nightly

### PR DESCRIPTION
**Pull request by geomin12**
## Motivation

as per discussion with team, the "phases" we will implement for enabling architectures will be:

stage 1: post-submit 
stage 2: pre-submit
stage 3: CI nightlies
stage 4: release pipeline

for these new phases, we need to have CI nightly run all archs and full tests

## Technical Details

Enabling all archs and full tests. 

Of course, we have limited machines, but enabling full tests will help us determine a good number for machine provisioning

## Test Plan

Verifying logic with Github Actions CI

## Test Result

Test results work [in setup step](https://github.com/ROCm/TheRock/actions/runs/19486785947/job/55770663975) and [CI summary](https://github.com/ROCm/TheRock/actions/runs/19486785947) (i forced workflow_dispatch to behave like is_schedule)

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: build_tools/github_actions/amdgpu_family_matrix.py [hunk 1]
```diff
@@ -191,8 +179,6 @@ def get_all_families_for_trigger_types(trigger_types):
     for trigger_type in trigger_types:
         if trigger_type in matrix_map:
             for family_name, family_config in matrix_map[trigger_type].items():
-                # Only add if not already present (first occurrence wins)
-                if family_name not in result:
-                    result[family_name] = family_config
+                result[family_name] = family_config
```
Comment (dezhiAmd) [inline build_tools/github_actions/amdgpu_family_matrix.py:204]: Concern - This is a significant behavioral change:
Previous behavior: First occurrence wins (presubmit overrides postsubmit overrides nightly)
New behavior: Last occurrence wins (nightly overrides postsubmit overrides presubmit)
Impact: When the same family appears in multiple matrices with different configs:
Before: The most restrictive config (presubmit) would be used
After: The least restrictive config (nightly) would be used
Example: For gfx110x:
Presubmit: "test-runs-on": "linux-gfx1101-gpu-rocm" with sanity_check_only_for_family: True
Nightly (before removal): "test-runs-on": "linux-gfx1101-gpu-rocm" with sanity_check_only_for_family: True
Since you removed gfx110x from nightly, this change might not affect anything now, but it's a breaking change for future configs.
Recommendation:
Either document this behavior change clearly, or
Keep the "first occurrence wins" logic and be explicit about matrix ordering
Consider using a merge strategy instead of overwrite (e.g., for test runners)

File: .github/workflows/ci_nightly.yml [hunk 2]
```diff
@@ -53,6 +53,8 @@ concurrency:
 jobs:
   setup:
     uses: ./.github/workflows/setup.yml
+    with:
+      build_variant: "release"
```
Comment (dezhiAmd) [inline .github/workflows/ci_nightly.yml:57]: Since it is nightly run, it would be nice to have asan build to test memory leak? For discussion

File: build_tools/github_actions/configure_ci.py [hunk 3]
```diff
@@ -382,10 +382,23 @@ def matrix_generator(
...
+        for key in amdgpu_family_info_matrix_all:
             selected_target_names.append(key)
 
+        # For nightly runs, we want to run full tests regardless of limited machines, so we delete the sanity_check_only_for_family option
+        for key in lookup_matrix:
+            if (
+                platform in lookup_matrix[key]
+                and "sanity_check_only_for_family" in lookup_matrix[key][platform]
+            ):
+                del lookup_matrix[key][platform]["sanity_check_only_for_family"]
```
Comment (dezhiAmd) [inline build_tools/github_actions/configure_ci.py:400]: Nit: Consider moving the comment added above to this line for code clarity

File: build_tools/github_actions/amdgpu_family_matrix.py (hunk 1 above)
Comment (geomin12) [reply build_tools/github_actions/amdgpu_family_matrix.py:204]: I noticed that `gfx110X` was duplicated for nightlies and presubmit. I removed it from nightlies and kept it in presubmit with the correct machines. This is why I removed the logic, though let me know if there is a need for it still

File: .github/workflows/ci_nightly.yml (hunk 2 above)
Comment (geomin12) [reply .github/workflows/ci_nightly.yml:57]: That is the eventual plan! ASAN is not working right now and working to fix. Once ASAN is working as expected, it will be added here

File: build_tools/github_actions/amdgpu_family_matrix.py (hunk 1 above)
Comment (dezhiAmd) [reply build_tools/github_actions/amdgpu_family_matrix.py:204]: I am not sure we want to enable all tests for pre-submit since that might take a long time.
But we definitely want to enable all tests for nightly CI to capture possible corner cases.

If above is true, then for same gfx110x, we need it appear in both nightly CI and pre-submit CI. But configurations is different.

**Comment by geomin12** [comment]
Test results work [in setup step](https://github.com/ROCm/TheRock/actions/runs/19486785947/job/55770663975) and [CI summary](https://github.com/ROCm/TheRock/actions/runs/19486785947) (i forced workflow_dispatch to behave like is_schedule)

Landing this
